[Bug]: Panic during checkpoint restart due to successor markets

### Problem encountered

I can see a panic in the checkpoint restart code related to successor markets if I run all successor markets tests from jake_successor_markets test branch **twice**. (Guessing one of the later tests does something that the LNL test earlier in the run doesn't like, but I'll try to narrow that down).

### Observed behaviour

Panic on second attempt

### Expected behaviour

No panics. Two successful runs back to back

### Steps to reproduce

```bash
1. pytest -m successor_markets
2. pytest -m successor_markets
```


### Software version

03b02dc9fed3ba9d24a912426345ad179d45fc79

### Failing test

_No response_

### Jenkins run

_No response_

### Configuration used

_No response_

### Relevant log output

```shell
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x2c88265]

goroutine 215 [running]:
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).Run.func1.1()
	/workspace/vega/cmd/vega/commands/node/node.go:137 +0x59
panic({0x48998a0, 0x7442c20})
	/usr/local/go/src/runtime/panic.go:884 +0x212
code.vegaprotocol.io/vega/core/governance.validateSuccessorMarket(0xc0035e2ec8, 0xc002e2ea60)
	/workspace/vega/core/governance/market.go:650 +0x85
code.vegaprotocol.io/vega/core/governance.validateNewMarketChange(0xc0035e2ec8, {0x57d14b0?, 0xc000244840}, 0x0?, {0x57d5388, 0xc00030df80}, 0x0?, 0xc002e2ecf0, 0x0?)
	/workspace/vega/core/governance/market.go:622 +0xcb
code.vegaprotocol.io/vega/core/governance.(*Engine).intoToSubmit(0xc000244f00, {0x57bdfc8, 0xc0012deba0}, 0xc0034e3600, 0xedc3f7daf?, 0x1)
	/workspace/vega/core/governance/engine.go:622 +0x318
code.vegaprotocol.io/vega/core/governance.(*Engine).Load(0xc000244f00, {0x57bdfc8, 0xc0012deba0}, {0xc00305c000, 0xabc4, 0xc000})
	/workspace/vega/core/governance/checkpoint.go:105 +0x799
code.vegaprotocol.io/vega/core/checkpoint.(*Engine).load(0xc00049ee10, {0x57bdfc8, 0xc0012deba0}, 0xc002e2f8f8)
	/workspace/vega/core/checkpoint/engine.go:337 +0xaf9
code.vegaprotocol.io/vega/core/checkpoint.(*Engine).UponGenesis(0xc00049ee10, {0x57bdfc8, 0xc0012deba0}, {0xc002124000, 0x1cde3, 0x1e000})
	/workspace/vega/core/checkpoint/engine.go:171 +0x4a8
code.vegaprotocol.io/vega/core/genesis.(*Handler).OnGenesis(0xc0015ca640, {0x57bdfc8, 0xc0012deba0}, {0x77555a0?, 0x46ae020?, 0x0?}, {0xc002124000, 0x1cde3, 0x1e000})
	/workspace/vega/core/genesis/handler.go:64 +0x3bd
code.vegaprotocol.io/vega/core/processor.(*App).OnInitChain(0xc000221200, {{0x19fa7d18, 0xedc3f7b2f, 0x0}, {0xc000623cd0, 0xb}, 0xc0010d2240, {0xc0012de9c0, 0x2, 0x2}, ...})
	/workspace/vega/core/processor/abci.go:708 +0x4ea
code.vegaprotocol.io/vega/core/blockchain/abci.(*App).InitChain(0xc0016cc160, {{0x19fa7d18, 0xedc3f7b2f, 0x0}, {0xc000623cd0, 0xb}, 0xc0010d2240, {0xc0012de9c0, 0x2, 0x2}, ...})
	/workspace/vega/core/blockchain/abci/abci.go:39 +0xbb
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*appW).InitChain(0x18?, {{0x19fa7d18, 0xedc3f7b2f, 0x0}, {0xc000623cd0, 0xb}, 0xc0010d2240, {0xc0012de9c0, 0x2, 0x2}, ...})
	/workspace/vega/cmd/vega/commands/node/app_wrapper.go:60 +0x7d
github.com/tendermint/tendermint/abci/client.(*localClient).InitChainSync(0xc00108aba0, {{0x19fa7d18, 0xedc3f7b2f, 0x0}, {0xc000623cd0, 0xb}, 0xc0010d2240, {0xc0012de9c0, 0x2, 0x2}, ...})
	/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/abci/client/local_client.go:272 +0x118
github.com/tendermint/tendermint/proxy.(*appConnConsensus).InitChainSync(0x1999971?, {{0x19fa7d18, 0xedc3f7b2f, 0x0}, {0xc000623cd0, 0xb}, 0xc0010d2240, {0xc0012de9c0, 0x2, 0x2}, ...})
	/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/proxy/app_conn.go:77 +0x55
github.com/tendermint/tendermint/consensus.(*Handshaker).ReplayBlocks(_, {{{0xb, 0x1}, {0x4ecaa39, 0x7}}, {0xc000623cd0, 0xb}, 0x1, 0x0, {{0x0, ...}, ...}, ...}, ...)
	/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/replay.go:319 +0xd78
github.com/tendermint/tendermint/consensus.(*Handshaker).Handshake(0xc002e30d90, {0x57ed218, 0xc000ef7c70})
	/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/replay.go:268 +0x3d4
github.com/tendermint/tendermint/node.doHandshake({_, _}, {{{0xb, 0x0}, {0x4ecaa39, 0x7}}, {0xc000623cd0, 0xb}, 0x1, 0x0, ...}, ...)
	/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/node/node.go:329 +0x1b8
github.com/tendermint/tendermint/node.NewNode(0xc00117f680, {0x57b5d60, 0xc0016c4640}, 0xc00105bf00, {0x5798680, 0xc000365578}, 0xc000079400?, 0x1?, 0xc00105bf40, {0x57bc470, ...}, ...)
	/home/jake-vega/go/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/node/node.go:779 +0x597
code.vegaprotocol.io/vega/core/blockchain/abci.NewTmNode({{0x0}, 0x1, 0x1, 0x0, 0x0, {0xc000f964b0, 0xa}, {{0x0}, {0xc000630228, 0x15}}, ...}, ...)
	/workspace/vega/core/blockchain/abci/tm_node.go:83 +0x705
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).startABCI(0xc000580800, 0x0?, {0x57e6b20, 0xc0004aca60}, {0x7fffd6df3f15, 0x2f}, {0x0?, 0x0?}, {0x0, 0x0})
	/workspace/vega/cmd/vega/commands/node/node.go:412 +0x1b6
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).startBlockchain(0xc000580800, 0x0?, {0x7fffd6df3f15?, 0x0?}, {0x0, 0x0}, {0x0, 0x0})
	/workspace/vega/cmd/vega/commands/node/node.go:299 +0x3ab
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).Run.func1()
	/workspace/vega/cmd/vega/commands/node/node.go:140 +0x90
created by code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).Run
	/workspace/vega/cmd/vega/commands/node/node.go:131 +0x66f
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -44,6 +44,7 @@
 - [8848](https://github.com/vegaprotocol/vega/issues/8848) - Handle the case where the market is terminated and the epoch ends at the same block.
 - [8853](https://github.com/vegaprotocol/vega/issues/8853) - Liquidity provision amendment bug fixes
 - [8862](https://github.com/vegaprotocol/vega/issues/8862) - Fix settlement via governance
+- [8772](https://github.com/vegaprotocol/vega/issues/8772) - Checkpoint panic on successor markets.
 
 
 ## 0.72.1

--- a/core/governance/engine.go
+++ b/core/governance/engine.go
@@ -602,18 +602,26 @@ func (e *Engine) intoToSubmit(ctx context.Context, p *types.Proposal, enct *enac
 		var parent *types.Market
 		if suc := newMarket.Successor(); suc != nil {
 			pm, ok := e.markets.GetMarket(suc.ParentID, true)
-			if !ok && !restore {
-				e.rejectProposal(ctx, p, types.ProposalErrorInvalidSuccessorMarket, ErrParentMarketDoesNotExist)
-				return nil, fmt.Errorf("%w, %v", ErrParentMarketDoesNotExist, types.ProposalErrorInvalidSuccessorMarket)
-			} else if restore && !ok {
-				newMarket.ClearSuccessor()
+			if !ok {
+				if !restore {
+					e.rejectProposal(ctx, p, types.ProposalErrorInvalidSuccessorMarket, ErrParentMarketDoesNotExist)
+					return nil, fmt.Errorf("%w, %v", ErrParentMarketDoesNotExist, types.ProposalErrorInvalidSuccessorMarket)
+				}
+			} else {
+				parent = &pm
 			}
 			// proposal to succeed a market that was already succeeded
-			if e.markets.IsSucceeded(suc.ParentID) {
+			// on restore, the parent market may be succeeded and the restored market may have own state (ie was the successor)
+			// So skip this check when restoring markets from checkpoints
+			if !restore && e.markets.IsSucceeded(suc.ParentID) {
 				e.rejectProposal(ctx, p, types.ProposalErrorInvalidSuccessorMarket, ErrParentMarketAlreadySucceeded)
 				return nil, fmt.Errorf("%w, %v", ErrParentMarketAlreadySucceeded, types.ProposalErrorInvalidSuccessorMarket)
 			}
-			parent = &pm
+			// CP restore market but the parent is not part of the state -> restore as regular market proposal
+			// in case the market has associated state
+			if restore && parent == nil {
+				newMarket.ClearSuccessor()
+			}
 		}
 		closeTime := time.Unix(p.Terms.ClosingTimestamp, 0)
 		enactTime := time.Unix(p.Terms.EnactmentTimestamp, 0)

--- a/core/governance/engine.go
+++ b/core/governance/engine.go
@@ -338,19 +338,10 @@ func (e *Engine) OnTick(ctx context.Context, t time.Time) ([]*ToEnact, []*VoteCl
 
 	now := t.Unix()
 
-	// use slice for deterministic behaviour and event order
-	// succeededMarkets := []string{}
-	// use map internally for O(1) lookups
-	ignoreSuccession := map[string]struct{}{}
 	for _, proposal := range e.activeProposals {
-		if proposal.IsSuccessorMarket() {
-			if _, ok := e.markets.GetMarket(proposal.ID, false); !ok {
-				// successor proposal for a successor market which cannot be enacted anymore -> remove the proposal
-				proposal.FailWithErr(types.ProposalErrorInvalidSuccessorMarket, ErrParentMarketAlreadySucceeded)
-				// ensure the event is sent
-				e.broker.Send(events.NewProposalEvent(ctx, *proposal.Proposal))
-			}
-		}
+		// do not check parent market, the market was either rejected when the parent was succeeded
+		// or, if the parent market state is gone (ie succession window has expired), the proposal simply
+		// loses its parent market reference
 		if proposal.ShouldClose(now) {
 			e.closeProposal(ctx, proposal)
 			voteClosed = append(voteClosed, e.preVoteClosedProposal(proposal))
@@ -360,19 +351,6 @@ func (e *Engine) OnTick(ctx context.Context, t time.Time) ([]*ToEnact, []*VoteCl
 			toBeRemoved = append(toBeRemoved, proposal.ID)
 		} else if proposal.IsPassed() && (e.isAutoEnactableProposal(proposal.Proposal) || proposal.IsTimeToEnact(now)) {
 			enact, perr, err := e.preEnactProposal(ctx, proposal)
-			if err == nil && proposal.IsSuccessorMarket() {
-				parentID, _ := proposal.NewMarket().ParentMarketID()
-				if _, ok := ignoreSuccession[parentID]; ok {
-					// @TODO we have a successor market ready, other proposals should not go through
-					// perhaps we ought to not set the errors here, but rather wait until
-					// we are sure the successor market went through
-					err = ErrParentMarketAlreadySucceeded
-					perr = types.ProposalErrorInvalidSuccessorMarket // @TODO proposal Error types
-				} else {
-					ignoreSuccession[parentID] = struct{}{}
-					// succeededMarkets = append(succeededMarkets, sucP.Changes.ParentID)
-				}
-			}
 			if err != nil {
 				e.broker.Send(events.NewProposalEvent(ctx, *proposal.Proposal))
 				toBeRemoved = append(toBeRemoved, proposal.ID)

